### PR TITLE
Companion for #13352

### PR DESCRIPTION
**Pull request by jsidorenko**
https://github.com/paritytech/substrate/pull/13352

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2382803

File: parachains/runtimes/assets/statemint/src/assets_api.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+use sp_std::vec::Vec;
+
+sp_api::decl_runtime_apis! {
+	pub trait AssetsApi<AccountId, AssetBalance, AssetId>
+	where
+		AccountId: Codec,
+		AssetBalance: Codec,
+		AssetId: Codec,
+	{
+		/// Returns the list of asset id and non-zero balances the AccountId has.
```
Comment (joepetrowski) [inline parachains/runtimes/assets/statemint/src/assets_api.rs:31]:
```suggestion
		/// Returns the list of `AssetId`s and corresponding balance that an `AccountId` has.
```

**Review by joepetrowski — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2383679

File: parachains/runtimes/assets/statemine/src/assets_api.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Runtime API definition for assets.
+
+use codec::Codec;
+use sp_std::vec::Vec;
+
+sp_api::decl_runtime_apis! {
+	pub trait AssetsApi<AccountId, AssetBalance, AssetId>
```
Comment (gilescope) [inline parachains/runtimes/assets/statemine/src/assets_api.rs:25]: @joepetrowski `AssetsApi` will become renamed to `LocalAssetsApi` and `ForeignAssetsApi` when https://github.com/paritytech/cumulus/pull/2133 lands?

Comment (bkontur) [reply parachains/runtimes/assets/statemine/src/assets_api.rs:25]: @joepetrowski @gilescope  @jsidorenko 

maybe it should/could be more general and we should have just one `AssetApi` which covers all pallet_assets instances

```
	pub trait AssetsApi<AccountId, MultiLocation>
	where
		AccountId: Codec,
	{
		/// Returns the list of `AssetId`s and corresponding balance that an `AccountId` has.
		fn account_balances(account: AccountId) -> Vec<(MultiLocation, u128)>;
	}
```

and implementation wil just do some mapping/concat
```
		fn account_balances(account: AccountId) -> Vec<(AssetIdForTrustBackedAssets, Balance)> {
			[
                Assets::account_balances(account).map(...ConvertedConcreteId<AssetIdForTrustBackedAssets..>::convert...)
                ForeignAssets::account_balances(account).map(...ConvertedConcreteId<MultiLocationForAssetId..>::convert...)
            ].concat()
		}
```

Comment (joepetrowski) [reply parachains/runtimes/assets/statemine/src/assets_api.rs:25]: Yeah I agree this is much friendlier to end users

File: parachains/runtimes/assets/statemint/src/assets_api.rs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Runtime API definition for assets.
+
+use codec::Codec;
+use sp_std::vec::Vec;
+
+sp_api::decl_runtime_apis! {
+	pub trait AssetsApi<AccountId, AssetBalance, AssetId>
```
Comment (bkontur) [inline parachains/runtimes/assets/statemint/src/assets_api.rs:25]: @joepetrowski [here](https://github.com/paritytech/cumulus/pull/2133#issue-1554990311) you mentioned `assets-common`, so I think this `AssetsApi` is also very hot candidate for that, can we start with `assets-common` here? I guess, this PR will go to the master before our ForeignAssets stuff

Comment (joepetrowski) [reply parachains/runtimes/assets/statemint/src/assets_api.rs:25]: Yeah I'm open to that, don't have time to refactor it myself in the near-term, and I would like to get 2133 in before the next release. But if someone can work on that and this fits in then by all means go for it.

**Comment by bkontur** [comment]
`assets-common` + `AssetsApi` refactor for multi pallet_assets instances is ready here: https://github.com/paritytech/cumulus/pull/2187

**Review by bkontur — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2395152

**Comment by gilescope** [comment]
There's definitely overlap with https://github.com/paritytech/cumulus/pull/2148 . I think 2148 should simplify some of this when it lands, but I don't think I have any objections to this PR as it stands. All roads seem to lead to AssetId being a multilocation.

**Comment by gilescope** [comment]
IS it possible that `AssetIdForTrustBackedAssetsConvert::reverse_ref(1)` could be made infallible? If so that would make things a tad simpler.

**Comment by bkontur** [comment]
> There's definitely overlap with #2148 . I think 2148 should simplify some of this when it lands, but I don't think I have any objections to this PR as it stands. All roads seem to lead to AssetId being a multilocation.

yes, agree, but maybe these two should be merged before #2148:
https://github.com/paritytech/cumulus/pull/2133
https://github.com/paritytech/cumulus/pull/2167

at least at #2167 - I have tests and working [AssetTransactor](https://github.com/paritytech/cumulus/pull/2167/files#diff-07287c62dd91150d975f7504bd14cc62d9889c44f7ba8ca44c30c03b05d5b071R141) with multiple instance:
```
pub type AssetTransactors = (CurrencyTransactor, FungiblesTransactor, ForeignFungiblesTransactor);
```

because there is a issue with `ConvertedConcreteId`, which I need to fix better: https://github.com/paritytech/cumulus/pull/2167/files#diff-aa9ec9164c4412c613444721bada9490dd697084f621658cba0a8db2ae360a13R19-R33

**Comment by bkontur** [comment]
> Is it possible that `AssetIdForTrustBackedAssetsConvert::reverse_ref(1)` could be made infallible? If so that would make things a tad simpler.

yes, it is not possible: https://github.com/paritytech/polkadot/blob/master/xcm/xcm-executor/src/traits/conversion.rs#L32-L47,
I wanted to reuse the existing conversions `AsPrefixedGeneralIndex`  - to use the same conversion everywhere,
thats why I needed to wrap that `Vec<()>` in `Result`

**Review by gilescope — APPROVED** [review]

**Comment by bkontur** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2422657

File: parachains/runtimes/assets/common/src/assets_api.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+sp_api::decl_runtime_apis! {
+	pub trait AssetsApi<AccountId, AssetBalance, AssetId>
+	where
+		AccountId: Codec,
+		AssetBalance: Codec,
+		AssetId: Codec,
+	{
+		/// Returns the list of `AssetId`s and corresponding balance that an `AccountId` has.
+		fn account_balances(account: AccountId) -> Result<Vec<(AssetId, AssetBalance)>, AssetsAccessError>;
```
Comment (kianenigma) [inline parachains/runtimes/assets/common/src/assets_api.rs:40]: Why is this re-defined?

Comment (bkchr) [reply parachains/runtimes/assets/common/src/assets_api.rs:40]: @bkontur you can just use MultiAsset in here as you can just import xcm stuff.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2426102

**Comment by bkontur** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2433325

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2434423

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2435437

**Comment by bkontur** [comment]
**Some context:**
iiuc, guys (working on DEX + UI) wanted to achieve with this companion and https://github.com/paritytech/substrate/pull/13352 :  **read account balances from pallet_assets from statemine/t**, I just try to help refactor here some common stuff, to be better prepared for multi pallet_assets instance (ForeignAssets).



**Actual status of this companion:**
- new runtime api `AssetsApi` defined here
   ```
   fn query_account_balances(account: AccountId) -> Result<Vec<xcm::latest::MultiAsset>, AssetsAccessError>   ```
- every runtime `statemine/statemint/westmint` implements this `AssetsApi`
  - collects `pallet_balance` + `pallet_assets (TrustBackedAssets instance)` and returns that as `Vec<MultiAsset>`
    -  (soon will be extended with `pallet_assets (ForeignAssets instance)`)
  - e.g. https://github.com/paritytech/cumulus/pull/2180/files#diff-84e2b90040060fcdc28fccc7e62463ecb9c01862b74efc71e84eeecef324184aR838-R857

---------------------

@kianenigma suggested [here](https://github.com/paritytech/substrate/pull/13352#pullrequestreview-1307305295) that it should be new api, xcm-based, and should be able to collect all MultiAssets from runtime (actually this new `AssetsApi` does exactly that)

@shawntabrizi mentioned [here](https://github.com/paritytech/substrate/pull/13352#issuecomment-1438665638) some "XCM Query RPC", so if we just move this `AssetsApi`to the polkadot xcm module somewhere, this could be a starting point and all parachains could use it.

yes, it is missing:
- pagination
- maybe some filter
- `max_iters: Option<u32>`

@KiChjang mentioned that there will be "XCM fungibles retreat ", so this is pretty good topic for that

---------------------


@joepetrowski @kianenigma @shawntabrizi 
so guys, please, somebody should tell what to do next:
- merge as it is https://github.com/paritytech/cumulus/pull/2180 + https://github.com/paritytech/substrate/pull/13352
- fix some minimal requirements a merge https://github.com/paritytech/cumulus/pull/2180 + https://github.com/paritytech/substrate/pull/13352
- postpone this

**Comment by joepetrowski** [comment]
> @joepetrowski @kianenigma @shawntabrizi
> so guys, please, somebody should tell what to do next:

I have discussed a bit with @kianenigma - the tentative plan now is that we merge these runtime RPCs as a short-term solution for better UI/UX development. But there will be someone joining soon who can work full time on a more holistic runtime interface for the long term. So I think we can proceed with this PR (but still need to address the above comment about making this part of Polkadot-XCM).

**Comment by bkontur** [comment]
> > @joepetrowski @kianenigma @shawntabrizi
> > so guys, please, somebody should tell what to do next:
> 
> I have discussed a bit with @kianenigma - the tentative plan now is that we merge these runtime RPCs as a short-term solution for better UI/UX development. But there will be someone joining soon who can work full time on a more holistic runtime interface for the long term. So I think we can proceed with this PR (but still need to address the above comment about making this part of Polkadot-XCM).

ok, cool, thank you,
I moved `AssetsApi` (renamed to `FungiblesApi`) to `xcm-builder` here: https://github.com/paritytech/polkadot/pull/6777,
once this PR is merged, I will change this companion to use it

**Comment by bkchr** [comment]
Please keep the runtime api here in this repository. It makes no sense to have them in Polkadot, especially if this will be some intermediate solution. People are also completely free to just copy this code, this isn't really complicated and there is also not that much code attached to this.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2450077

**Comment by bkontur** [comment]
> Please keep the runtime api here in this repository. It makes no sense to have them in Polkadot, especially if this will be some intermediate solution. People are also completely free to just copy this code, this isn't really complicated and there is also not that much code attached to this.

thank you, makes sense,
I refactored that `runtime_api.rs` definition, now is clean and the same as file from [polkadot/xcm PR](https://github.com/paritytech/polkadot/pull/6777/files#diff-df0fea42045207170f8b159d2640ccfcd00da630bd57f38a2cca8a9ee79f698b)

if we get one more approve here, I can close that polkadot/xcm PR (unless anybody really wants to put it to the polkadot/xcm)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2450271

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2450410

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2453937

File: parachains/runtimes/assets/common/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+# Polkadot
+xcm = { git = "https://github.com/paritytech/polkadot", default-features = false, branch = "master" }
+xcm-builder = { git = "https://github.com/paritytech/polkadot", default-features = false, branch = "master" }
+xcm-executor = { git = "https://github.com/paritytech/polkadot", default-features = false, branch = "master" }
+
+# Cumulus
+parachains-common = { path = "../../../common", default-features = false }
+
+#[dev-dependencies]
```
Comment (bkchr) [inline parachains/runtimes/assets/common/Cargo.toml:24]:
```suggestion
```

**Review by bkchr — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2472725

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2473890

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v0-9-40/2468/1

**Pull request merged**